### ISSUE TITLE
Associated types aren't normalized with impl cycles

### ISSUE THREAD

**Issue by ubsan** [rust-lang/rust#31593]
``` rust
trait Doop {
    type Shaboom;
    fn badoowop(self) -> Self::Shaboom;
}
trait BaDoop: Doop { }

impl<T: BaDoop> Doop for T {
    type Shaboom = T;
    fn badoowop(self) -> Self {
        self
    }
}

fn main() { }
```

errors with

```
error: method `badoowop` has an incompatible type for trait: 
 expected associated type,
 found type parameter
```

[playground](http://is.gd/FnD5wQ)

A similar issue happens when you now return `Self::Shaboom`:

```
error: mismatched types: expected `<T as Doop>::Shaboom`,
    found `T`
```

[playground](http://is.gd/1BNlTP)

**Comment by ubsan** [comment]
Honestly, it probably just needs a better error message.

**Comment by eddyb** [comment]
This seems related to #29859, but not... actually working?
AFAICT, `<T as Doop>` in the `impl<T: Doop> of Doop for T` cannot refer to the same impl.

cc @nikomatsakis

**Comment by ubsan** [comment]
@eddyb looks like the forest fixed this? #30533

**Comment by oli-obk** [comment]
The error message seems reasonable to me now as we get

```
error[E0053]: method `badoowop` has an incompatible type for trait
 --> src/main.rs:9:26
  |
3 |     fn badoowop(self) -> Self::Shaboom;
  |                          ------------- type in trait
...
9 |     fn badoowop(self) -> Self {
  |                          ^^^^ expected associated type, found type parameter
  |
  = note: expected type `fn(T) -> <T as Doop>::Shaboom`
             found type `fn(T) -> T`
  = note: you might be missing a type parameter or trait bound
```

Please reopen if you disagree
